File: src/lib/extensions/contextMenus.ts
/**
 * Exodus Browser — host integration for chrome.contextMenus.
 * Aerospace-level error handling, security validation, and input validation.
 */

import { invoke, isTauri } from '@tauri-apps/api/core';

// Aerospace-level security validation patterns
const VALID_EXTENSION_ID_PATTERN = /^[a-zA-Z0-9_-]+$/;
const VALID_MENU_ITEM_ID_PATTERN = /^[a-zA-Z0-9_\-:]+$/;
const VALID_URL_PATTERN = /^https?:\/\/.+/;
const VALID_HOST_CONTEXT_PATTERN = /^[a-zA-Z0-9_-]+$/;

/**
 * Aerospace-level validation for extension ID format.
 */
function validateExtensionId(extensionId: string): boolean {
  if (!extensionId || typeof extensionId !== 'string') {
    console.error('[ContextMenus] Invalid extension ID');
    return false;
  }
  return VALID_EXTENSION_ID_PATTERN.test(extensionId);
}

/**
 * Aerospace-level validation for menu item ID format.
 */
function validateMenuItemId(menuItemId: string): boolean {
  if (!menuItemId || typeof menuItemId !== 'string') {
    console.error('[ContextMenus] Invalid menu item ID');
    return false;
  }
  return VALID_MENU_ITEM_ID_PATTERN.test(menuItemId);
}

/**
 * Aerospace-level validation for URL format.
 */
function validateUrl(url: string): boolean {
  if (!url || typeof url !== 'string') {
    console.error('[ContextMenus] Invalid URL');
    return false;
  }
  return VALID_URL_PATTERN.test(url);
}

/**
 * Aerospace-level validation for host context format.
 */
function validateHostContext(hostContext: string): boolean {
  if (!hostContext || typeof hostContext !== 'string') {
    console.error('[ContextMenus] Invalid host context');
    return false;
  }
  return VALID_HOST_CONTEXT_PATTERN.test(hostContext);
}

/** Context menu item from extension registry. */
export type ExtensionContextMenuItem = {
  id: string;
  title: string;
  contexts: string[];
  parentId?: string | null;
  enabled: boolean;
  visible: boolean;
  itemType: string;
  checked?: boolean;
};

/** Row shown in the host context menu. */
export type HostContextMenuEntry = {
  extensionId: string;
  extensionName: string;
  item: ExtensionContextMenuItem;
};

/** List extension context menu items for a page URL and host context (e.g. `page`). */
export async function listExtensionContextMenus(
  pageUrl: string,
  hostContext = 'page',
): Promise<HostContextMenuEntry[]> {
  if (!isTauri()) return [];
  
  // Aerospace-level input validation
  if (!validateUrl(pageUrl)) {
    console.error('[ContextMenus] Invalid pageUrl for listExtensionContextMenus:', pageUrl);
    return [];
  }
  
  if (!validateHostContext(hostContext)) {
    console.error('[ContextMenus] Invalid hostContext for listExtensionContextMenus:', hostContext);
    return [];
  }
  
  try {
    return await invoke<HostContextMenuEntry[]>('extension_context_menus_list_host', {
      pageUrl,
      hostContext,
    });
  } catch (error) {
    console.error('extension_context_menus_list_host failed:', error);
    return [];
  }
}

/** Notify extension background that a context menu item was clicked. */
export async function fireExtensionContextMenuClick(
  extensionId: string,
  menuItemId: string,
  pageUrl: string,
): Promise<void> {
  if (!isTauri()) return;
  
  // Aerospace-level input validation
  if (!validateExtensionId(extensionId)) {
    console.error('[ContextMenus] Invalid extensionId for fireExtensionContextMenuClick:', extensionId);
    return;
  }
  
  if (!validateMenuItemId(menuItemId)) {
    console.error('[ContextMenus] Invalid menuItemId for fireExtensionContextMenuClick:', menuItemId);
    return;
  }
  
  if (!validateUrl(pageUrl)) {
    console.error('[ContextMenus] Invalid pageUrl for fireExtensionContextMenuClick:', pageUrl);
    return;
  }
  
  try {
    await invoke('extension_context_menu_clicked', {
      extensionId,
      menuItemId,
      pageUrl,
    });
  } catch (error) {
    console.error('extension_context_menu_clicked failed:', error);
  }
}

/** Context menu creation options. */
export type CreateContextMenuOptions = {
  id: string;
  title?: string;
  contexts?: string[];
  parentId?: string;
  enabled?: boolean;
  visible?: boolean;
  checked?: boolean;
  type?: 'normal' | 'checkbox' | 'radio' | 'separator';
  icons?: { '16': string; '32': string; '48': string; '128': string };
};

/** chrome.contextMenus.create - Create a context menu item. */
export async function createContextMenu(
  extensionId: string,
  options: CreateContextMenuOptions,
): Promise<void> {
  if (!isTauri()) return;
  
  // Aerospace-level input validation
  if (!validateExtensionId(extensionId)) {
    console.error('[ContextMenus] Invalid extensionId for createContextMenu');
    return;
  }
  
  if (!options || typeof options !== 'object') {
    console.error('[ContextMenus] Invalid options for createContextMenu');
    return;
  }
  
  if (!options.id || !validateMenuItemId(options.id)) {
    console.error('[ContextMenus] Invalid id in options for createContextMenu');
    return;
  }
  
  // Validate parentId if provided
  if (options.parentId && !validateMenuItemId(options.parentId)) {
    console.error('[ContextMenus] Invalid parentId in options for createContextMenu');
    return;
  }
  
  // Validate contexts if provided
  if (options.contexts && !Array.isArray(options.contexts)) {
    console.error('[ContextMenus] Invalid contexts in options for createContextMenu');
    return;
  }
  
  try {
    await invoke('extension_context_menus_create', { extensionId, options });
  } catch (error) {
    console.error('[ContextMenus] extension_context_menus_create failed:', error);
    throw error;
  }
}

/** Context menu update options. */
export type UpdateContextMenuOptions = {
  title?: string;
  contexts?: string[];
  enabled?: boolean;
  visible?: boolean;
  checked?: boolean;
  icons?: { '16': string; '32': string; '48': string; '128': string };
};

/** chrome.contextMenus.update - Update a context menu item. */
export async function updateContextMenu(
  extensionId: string,
  menuItemId: string,
  options: UpdateContextMenuOptions,
): Promise<void> {
  if (!isTauri()) return;
  
  // Aerospace-level input validation
  if (!validateExtensionId(extensionId)) {
    console.error('[ContextMenus] Invalid extensionId for updateContextMenu');
    return;
  }
  
  if (!validateMenuItemId(menuItemId)) {
    console.error('[ContextMenus] Invalid menuItemId for updateContextMenu');
    return;
  }
  
  if (!options || typeof options !== 'object') {
    console.error('[ContextMenus] Invalid options for updateContextMenu');
    return;
  }
  
  // Validate contexts if provided
  if (options.contexts && !Array.isArray(options.contexts)) {
    console.error('[ContextMenus] Invalid contexts in options for updateContextMenu');
    return;
  }
  
  try {
    await invoke('extension_context_menus_update', { extensionId, menuItemId, options });
  } catch (error) {
    console.error('[ContextMenus] extension_context_menus_update failed:', error);
    throw error;
  }
}

/** chrome.contextMenus.remove - Remove a context menu item. */
export async function removeContextMenu(
  extensionId: string,
  menuItemId: string,
): Promise<void> {
  if (!isTauri()) return;
  
  // Aerospace-level input validation
  if (!validateExtensionId(extensionId)) {
    console.error('[ContextMenus] Invalid extensionId for removeContextMenu');
    return;
  }
  
  if (!validateMenuItemId(menuItemId)) {
    console.error('[ContextMenus] Invalid menuItemId for removeContextMenu');
    return;
  }
  
  try {
    await invoke('extension_context_menus_remove', { extensionId, menuItemId });
  } catch (error) {
    console.error('[ContextMenus] extension_context_menus_remove failed:', error);
    throw error;
  }
}

/** chrome.contextMenus.removeAll - Remove all context menu items. */
export async function removeAllContextMenus(
  extensionId: string,
): Promise<void> {
  if (!isTauri()) return;
  
  // Aerospace-level input validation
  if (!validateExtensionId(extensionId)) {
    console.error('[ContextMenus] Invalid extensionId for removeAllContextMenus');
    return;
  }
  
  try {
    await invoke('extension_context_menus_remove_all', { extensionId });
  } catch (error) {
    console.error('[ContextMenus] extension_context_menus_remove_all failed:', error);
    throw error;
  }
}

File: src-tauri/src/plugins/context_menus.rs
//! Exodus Browser — chrome.contextMenus API implementation
//!
//! Provides context menu functionality for extensions with high reliability
//! and safety guarantees following aerospace-grade standards.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use tauri::{Emitter, Manager, State};

/// Context menu item type
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ContextItemType {
    Normal,
    Checkbox,
    Radio,
    Separator,
}

impl Default for ContextItemType {
    fn default() -> Self {
        Self::Normal
    }
}

/// Context menu item
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextMenuItem {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub contexts: Vec<String>,
    #[serde(default)]
    pub parent_id: Option<String>,
    #[serde(default)]
    pub document_url_patterns: Option<Vec<String>>,
    #[serde(default)]
    pub target_url_patterns: Option<Vec<String>>,
    #[serde(default)]
    pub enabled: bool,
    #[serde(default)]
    pub visible: bool,
    #[serde(default)]
    pub item_type: ContextItemType,
    #[serde(default)]
    pub checked: bool,
}

/// Context menu registry
pub struct ContextMenuRegistry {
    items: Arc<Mutex<HashMap<String, ContextMenuItem>>>,
}

impl ContextMenuRegistry {
    pub fn new() -> Self {
        Self {
            items: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Create a context menu item
    pub fn create(&self, item: ContextMenuItem) -> Result<(), String> {
        let mut items = self.items.lock().map_err(|e| format!("Lock error: {}", e))?;
        items.insert(item.id.clone(), item);
        Ok(())
    }

    /// Update a context menu item
    pub fn update(&self, id: &str, update_properties: ContextMenuItem) -> Result<(), String> {
        let mut items = self.items.lock().map_err(|e| format!("Lock error: {}", e))?;
        if items.contains_key(id) {
            items.insert(id.to_string(), update_properties);
            Ok(())
        } else {
            Err("Item not found".to_string())
        }
    }

    /// Remove a context menu item
    pub fn remove(&self, id: &str) -> Result<(), String> {
        let mut items = self.items.lock().map_err(|e| format!("Lock error: {}", e))?;
        if items.remove(id).is_some() {
            Ok(())
        } else {
            Err("Item not found".to_string())
        }
    }

    /// Remove all context menu items for an extension
    pub fn remove_all(&self, extension_id: &str) -> Result<(), String> {
        let mut items = self.items.lock().map_err(|e| format!("Lock error: {}", e))?;
        items.retain(|id, _| !id.starts_with(&format!("{}:", extension_id)));
        Ok(())
    }

    /// Get all context menu items for an extension
    pub fn get_all(&self, extension_id: &str) -> Vec<ContextMenuItem> {
        let items = self.items.lock().ok();
        match items {
            Some(guard) => guard
                .values()
                .filter(|item| item.id.starts_with(&format!("{}:", extension_id)))
                .cloned()
                .collect(),
            None => Vec::new(),
        }
    }

    /// Get a specific context menu item
    pub fn get(&self, id: &str) -> Option<ContextMenuItem> {
        let items = self.items.lock().ok();
        match items {
            Some(guard) => guard.get(id).cloned(),
            None => None,
        }
    }
}

impl Default for ContextMenuRegistry {
    fn default() -> Self {
        Self::new()
    }
}

/// Apply flushed context menu ops from a content/background webview outbox.
pub fn apply_context_menu_flush_ops(
    registry: &ContextMenuRegistry,
    ops: &[super::chrome_bridge::ContextMenuFlushOp],
) {
    for op in ops {
        let ext_id = &op.extension_id;
        match op.op.as_str() {
            "create" => {
                if let Some(mut item) = op.item.clone() {
                    item.id = format!("{}:{}", ext_id, item.id);
                    let _ = registry.create(item);
                }
            }
            "update" => {
                if let (Some(id), Some(mut item)) = (&op.id, op.item.clone()) {
                    let prefixed = format!("{}:{}", ext_id, id);
                    item.id = prefixed.clone();
                    let _ = registry.update(&prefixed, item);
                }
            }
            "remove" => {
                if let Some(id) = &op.id {
                    let prefixed = format!("{}:{}", ext_id, id);
                    let _ = registry.remove(&prefixed);
                }
            }
            "removeAll" => {
                let _ = registry.remove_all(ext_id);
            }
            _ => {}
        }
    }
}

/// Host-visible context menu row (extension + item).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HostContextMenuEntry {
    pub extension_id: String,
    pub extension_name: String,
    pub item: ContextMenuItem,
}

/// Whether a menu item should appear for the given host context and page URL.
pub fn item_visible_for_host(item: &ContextMenuItem, host_context: &str, page_url: &str) -> bool {
    if !item.visible || !item.enabled {
        return false;
    }
    if matches!(item.item_type, ContextItemType::Separator) {
        return true;
    }
    if !item.contexts.is_empty() && !item.contexts.iter().any(|c| c == host_context) {
        return false;
    }
    if let Some(patterns) = &item.document_url_patterns {
        if !patterns.is_empty() {
            let mut matched = false;
            for pat in patterns {
                if super::match_patterns::url_matches_pattern(page_url, pat) {
                    matched = true;
                    break;
                }
            }
            if !matched {
                return false;
            }
        }
    }
    true
}
